src/transport/manager/handle.rs

```diff
@@ -180,50 +180,66 @@ impl TransportManagerHandle {
         peer: &PeerId,
         addresses: impl Iterator<Item = Multiaddr>,
     ) -> usize {
-        let mut peers = self.peers.write();
-        let addresses = addresses
-            .filter_map(|address| {
-                (self.supported_transport(&address) && !self.is_local_address(&address))
-                    .then_some(AddressRecord::from_multiaddr(address)?)
-            })
-            .collect::<HashSet<_>>();
-
-        // if all of the added addresses belonged to unsupported transports, exit early
-        let num_added = addresses.len();
-        if num_added == 0 {
-            tracing::debug!(
-                target: LOG_TARGET,
-                ?peer,
-                "didn't add any addresses for peer because transport is not supported",
-            );
+        let mut peer_addresses = HashMap::new();
+
+        for address in addresses {
+            // There is not supported transport configured that can dial this address.
+            if !self.supported_transport(&address) {
+                continue;
+            }
+            if self.is_local_address(&address) {
+                continue;
+            }
+
+            // Check the peer ID if present.
+            if let Some(Protocol::P2p(multihash)) = address.iter().last() {
+                // Ignore the address if the peer ID is invalid.
+                let Ok(peer_id) = PeerId::from_multihash(multihash.clone()) else {
+                    continue;
+                };
+
+                // This can correspond to the provided peerID or to a different one.
+                if peer_id != *peer {
+                    tracing::warn!(
+                        target: LOG_TARGET,
+                        ?peer,
+                        ?address,
+                        "Added known address that corresponds to a different peer ID",
+                    );
+                }
 
-            return 0usize;
+                // It is important to keep track of all addresses to have a healthy
+                // address store to dial from.
+                peer_addresses.entry(peer_id).or_insert_with(HashSet::new).insert(address);
```

Comment:
Why not discard the address with invalid peer ID? If the address with invalid peer ID ended up here, something bad has already happened. In theory, we could rewrite the peer ID part, but ideally we should have returned an error to the user long before coming here with invalid peer ID.

This would also simplify the code below by inserting only one peer in the map.

Reply:
Oki doki that makes sense. I think I'm convinced that tracking all peer addresses is not the best thing here, since the user explicitly provided addresses of a single "peer". I've discarded addresses that belong to a different peer,this should make the code a bit easier to follow, thanks for the suggestion 🙏